### PR TITLE
Module 7

### PR DESCRIPTION
**Pull request by debasen**
# Requirement
## Install a Form Library

Currently, the most popular form libraries for React are React Hook Form and Formik. We recommend reviewing the documentation for each and then selecting the one that best fits your needs. Consult with your mentor when making this decision.

## Implement a Possibility to Add Movies

In the Advanced Components module, you should have created a Storybook story showcasing a composition of Dialog and MovieForm. In this module, extend this logic to have a fully functioning form that allows adding new movies.

If you open the design prototype, you'll notice that the form is placed in a dialog window on top of the movie list page. To achieve this behavior without losing context when adding a new movie, define another child route to the route that renders the `<SearchForm />` element. Use "/new" as the value for the path and render a new `AddMovieForm` component (you need to create it). As a result, you should have two levels of nesting. Then add an `<Outlet />` to the `SearchForm` component so that the router can render nested routes together with the movie search form.

Now, if you open the app and navigate to "/new", you should be able to see the movie list page with the movie search form on top and the contents of your new `AddMovieForm` component. In the component, render your composition of Dialog + MovieForm so that your form appears in a dialog on top of the existing screen.

Extend the `MovieForm` component to use your form library of choice. Add form fields validation. In the `AddMovieForm`, handle form submission and send a request to an API endpoint for adding new movies. When the request fulfills, close the modal by navigating the user to "/". Alternatively, you can navigate them to /:movieId, where :movieId is the ID of the newly created movie, obtained from the server response.

On the design prototype, there's a link to add a new movie. Implement it so that when a user clicks it, the route changes to "/new". Optionally, you can preserve the current search query parameters so that the movie list is not updated when you click on the link.

By the end of this step, you should have the possibility to add new movies and then search for them in the list using the previously built search functionality.

## Implement a Possibility to Edit a Movie

To edit an existing movie, you will need another route. According to the design prototype, editing also happens in a dialog window, similar to adding a new movie. Think about the best place to create your child route and create it similar to adding a new movie. Define the pathname yourself, although we recommend "/:movieId/edit".

Similar to adding a new movie, implement functionality to edit an existing movie. The difference is that you will need to load movie details and then send them to the `MovieForm`, so that all form fields get populated with existing data.

In a `MovieCard` context menu, there's a link to edit the movie. When a user clicks on the link, navigate them to "/:movieId/edit" or whatever pathname you have defined.

By the end of this step, you should have the possibility to edit any movie from the list.

## Write End-to-End Tests

Cover the newly added functionality with more end-to-end tests. Verify that navigating the "Add Movie" link opens a dialog with an empty movie form, and selecting "Edit" from a movie context menu opens a dialog with pre-populated movie details. Write a test that adds a new movie and verify it appears in the list after.

# Verify
## Clips
[edit-scenario.webm](https://github.com/debasen/netfilx-roulette/assets/28741376/c2b2d11c-0a0a-4844-b17b-9ab2c3edd716)
[add-scenario.webm](https://github.com/debasen/netfilx-roulette/assets/28741376/8e49c6c2-b014-49ad-bdf1-43cdfacb2cd3)

##Screenshots
### Add Movie
image
### Edit Movie
image

## E2E

image

### CODE REVIEW

File: src/components/movieForm/movieForm.js [hunk 1]
```diff
@@ -89,55 +126,71 @@ export const MovieForm = ({ movie, onSubmit }) => {
               className="form-input form-placeholder"
               type="text"
               id="movieUrl"
-              name="movieUrl"
+              {...register('movieUrl', {
+                pattern: { value: /^(http|https):\/\/[^ "]+$/, message: 'Not a valid URL' },
+                required: 'Image URL is required'
+              })}
```
Comment (melvinvicky46) [inline src/components/movieForm/movieForm.js]: Write this code in a separate method and call it over here.

File: src/components/movieForm/movieForm.js [hunk 2]
```diff
@@ -89,55 +126,71 @@ export const MovieForm = ({ movie, onSubmit }) => {
...
               id="rating"
-              name="rating"
+              step={.1}
+              {...register('rating',
+                {
+                  required: { value: true, message: 'Rating is required' },
+                  max: { value: 10, message: 'Allowed value is 0-10' },
+                  min: { value: 0, message: 'Allowed value is 0-10' }
+                })
+              }
```
Comment (melvinvicky46) [inline src/components/movieForm/movieForm.js]: Write this code in a separate method and call it over here.

File: src/components/movieForm/movieForm.js [hunk 3]
```diff
@@ -89,55 +126,71 @@ export const MovieForm = ({ movie, onSubmit }) => {
...
+            <Select
+              classNamePrefix='form-select'
+              {...register('genres', { required: true })} // Use register properly
+              options={[
+                { value: 'Action', label: 'Action' },
+                { value: 'Comedy', label: 'Comedy' },
+                { value: 'Drama', label: 'Drama' },
+                { value: 'Horror', label: 'Horror' },
+                { value: 'Science Fiction', label: 'Science Fiction' },
+              ]}
```
Comment (melvinvicky46) [inline src/components/movieForm/movieForm.js]: Declare an const and use it.

File: src/components/movieForm/movieForm.js [hunk 4]
```diff
@@ -89,55 +126,71 @@ export const MovieForm = ({ movie, onSubmit }) => {
...
               className="form-input form-placeholder"
-              type="text"
+              type="number"
               id="runtime"
-              name="runtime"
+              step={1}
+              {...register('runtime', { required: 'Runtime is required' },
+                { min: 1, message: 'Allowed value is 0-1000' },
+                { max: 1000, message: 'Allowed value is 0-1000' })
+              }
```
Comment (melvinvicky46) [inline src/components/movieForm/movieForm.js]: Write this code in a separate method and call it over here.

File: src/components/movieForm/movieForm.scss [hunk 5]
```diff
@@ -59,7 +59,37 @@
...
+    }
+
+    /*Remove arrows beside number field*/
+    input[type=number]::-webkit-inner-spin-button,
+    input[type=number]::-webkit-outer-spin-button {
+        -webkit-appearance: none;
+        -moz-appearance: none;
+        appearance: none;
+        margin: 0;
+    }
```
Comment (melvinvicky46) [inline src/components/movieForm/movieForm.scss]: write a custom className and nest the built-in className if any.

**Review by melvinvicky46 — COMMENTED** [review]
Please update the changes.

File: src/components/movieForm/movieForm.js (hunk 1 above)
Comment (debasen) [reply src/components/movieForm/movieForm.js]: Created separate method.

File: src/components/movieForm/movieForm.js (hunk 2 above)
Comment (debasen) [reply src/components/movieForm/movieForm.js]: Created separate method.

File: src/components/movieForm/movieForm.js (hunk 3 above)
Comment (debasen) [reply src/components/movieForm/movieForm.js]: Moved to a constant

File: src/components/movieForm/movieForm.js (hunk 4 above)
Comment (debasen) [reply src/components/movieForm/movieForm.js]: Created separate method.

File: src/components/movieForm/movieForm.scss (hunk 5 above)
Comment (debasen) [reply src/components/movieForm/movieForm.scss]: Created a custom class and moved to nested class

**Review by melvinvicky46 — APPROVED** [review]

**Pull request merged**